mapstructure.go

```diff
@@ -0,0 +1,45 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+// StringToFileMode returns a function that converts strings to os.FileMode
+// value. This is designed to be used with mapstructure for parsing out a
+// filemode value.
+func StringToFileMode() mapstructure.DecodeHookFunc {
+	return func(
+		f reflect.Type,
+		t reflect.Type,
+		data interface{}) (interface{}, error) {
+		if f.Kind() != reflect.String {
+			return data, nil
+		}
+		if t != reflect.TypeOf(os.FileMode(0)) {
+			return data, nil
+		}
+
+		typed := data.(string)
+		switch len(typed) {
+		case 3:
+			// Assume the user didn't set the special bit. Some users just put perms
+			// as "600", but they really meant "0600".
+			typed = string(append([]byte("0"), []byte(typed)...))
+		case 4:
+		default:
+			return data, fmt.Errorf("invalid filemode format: %q", typed)
+		}
+
+		// Convert it by parsing
+		v, err := strconv.ParseUint(typed, 8, 32)
```

Comment:
Can you change this to `strconv.ParseUint(typed, 8, 12)` instead?  We don't need the other 20 bits when parsing the mode.  If you do that, you can also pass `data` directly to `ParseUint` without prepending a leading "0" because `ParseUint` will do the right thing.